[JENKINS-31661] Check the root url (Admin monitor + Wizard)

Wadeck opened the pull request:
The problem is that the application has some little issues when there is no rootURL configured. Example: [JENKINS-47426](https://issues.jenkins-ci.org/browse/JENKINS-47426).
Normally with the Setup Wizard, this should not be the case. But there are some exceptions like development instance, administrator that disabled the wizard or if the administrator simply clear the root URL in the config.

There was already a PR ([#1921](https://github.com/jenkinsci/jenkins/pull/1921)) for the issue but as the PR went too far in the correction it was not merged (as I understand). This PR focuses only on the monitor part.

See [JENKINS-31661](https://issues.jenkins-ci.org/browse/JENKINS-31661).

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* Add administrative monitor to check the configuration of the root URL
* Add a new page in the Wizard to let the administrator configure the root URL
* Add an additional option in `HttpResponses` for sending an error.

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@reviewbybees @oleg-nenashev @amuniz

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Screenshot
<img width="485" alt="root_url_not_set_monitor_message" src="https://user-images.githubusercontent.com/2662497/31554908-66d83274-b03f-11e7-80d3-171134422179.PNG">

Wadeck committed:
[JENKINS-31661] Add an administrative monitor to check the root url configuration
- add the monitor + test

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

oleg-nenashev commented on core/src/main/java/jenkins/diagnostics/RootUrlNotSetMonitor.java:
```diff
@@ -0,0 +1,54 @@
+ * Normally this root URL is set during SetupWizard phase, this monitor is there to ensure that behavior.
+ * Potential exceptions are the dev environment, if someone disable the wizard or
+ * the administrator put an empty string on the configuration page.
+ */
```
`@since` may be useful here though not strictly required

Wadeck replied:
added

oleg-nenashev commented on core/src/main/resources/jenkins/diagnostics/Messages.properties:
```diff
@@ -1,3 +1,4 @@
 CompletedInitializationMonitor.DisplayName=Jenkins Initialization Monitor
 SecurityIsOffMonitor.DisplayName=Disabled Security
 URICheckEncodingMonitor.DisplayName=Check URI Encoding
+RootUrlNotSetMonitor.DisplayName=Root URL configured Monitor
```
"not" is missing somewhere, I'd guess

Wadeck replied:
added

oleg-nenashev commented on test/src/test/java/jenkins/diagnostics/RootUrlNotSetMonitorTest.java:
```diff
@@ -0,0 +1,59 @@
+        assertTrue(StringUtils.isNotBlank(JenkinsLocationConfiguration.get().getUrl()));
+        
+        RootUrlNotSetMonitor monitor = j.jenkins.getExtensionList(AdministrativeMonitor.class).get(RootUrlNotSetMonitor.class);
+        assertFalse("Monitor must be activated", monitor.isActivated());
```
"not be activated"

Wadeck replied:
corrected

oleg-nenashev commented on test/src/test/java/jenkins/diagnostics/RootUrlNotSetMonitorTest.java:
```diff
@@ -0,0 +1,59 @@
+        JenkinsLocationConfiguration.get().setUrl(null);
+        
+        RootUrlNotSetMonitor monitor = j.jenkins.getExtensionList(AdministrativeMonitor.class).get(RootUrlNotSetMonitor.class);
+        assertTrue("Monitor must not be activated", monitor.isActivated());
```
"must be activated"

Wadeck replied:
corrected

oleg-nenashev commented on test/src/test/java/jenkins/diagnostics/RootUrlNotSetMonitorTest.java:
```diff
@@ -0,0 +1,59 @@
+    @Test
+    @Issue("JENKINS-31661")
+    public void testWithRootUrl_configured() {
+        assertTrue(StringUtils.isNotBlank(JenkinsLocationConfiguration.get().getUrl()));
```
The test relies on the default JTH behavior. Maybe it makes sense to mention it

Wadeck replied:
good point :)

oleg-nenashev reviewed (changes requested):
🐛 The error messages are inversed in the tests. It should be fixed. The rest is fine IMHO.

Wadeck committed:
- corrections requested by Oleg in code review

oleg-nenashev reviewed (dismissed):
🐝

oleg-nenashev commented:
@reviewbybees done

daniel-beck commented:
I just tried with the PR build: Going through the setup wizard leaves Jenkins without a defined root URL. 

While I think this admin monitor is a good idea, Jenkins should not be showing warnings (other than "newer Jenkins is available") to someone who just went through the setup wizard.

For that reason, I strongly disagree with merging this change as is. The setup wizard should be extended to allow defining a Jenkins URL (or could perhaps infer one from the URL used to access Jenkins in the setup wizard, similar to how the form field determines its default value?).

Only then would this change make sense.

daniel-beck reviewed (dismissed):
See previous comment; this does not make sense if the setup wizard leaves the root URL undefined.

Wadeck committed:
- add a field to fill during SetupWizard for the rootUrl (default value using javascript but editable)
- WIP : missing the translation

Wadeck commented:
Werk in progress for the modification in the SetupWizard flow. Under the panel which contain the admin user creation form, I added the new field for RootURL.
Screenshot is coming...

Wadeck committed:
- revert whitespace change

Wadeck committed:
- little refactoring of js methods

Wadeck committed:
- translation added

Wadeck commented:
@kzantow do you have time to review those modifications ?
@daniel-beck  as requested, the field is proposed during setup, so if you finish the Setup, the root URL will be set.

oleg-nenashev commented:
@Wadeck After "werk in progress" I rather expected to see a German translation :D

oleg-nenashev commented:
I am a bit aware about the "Proceed as current admin" hyperlink. If I click this button, will the local URL change be applied?

P.S: I would be happy to remove this button at all. It causes too many issues with unintended clicks

Wadeck commented:
@oleg-nenashev Yes, I put my "save action" for both the skip and save first user. And as it's a button and not a link, no risk to see somebody doing some ctrl+click or right-click open in a new tab, etc. So _normally_ it should be called all the time.

jglick commented on core/src/main/java/jenkins/diagnostics/RootUrlNotSetMonitor.java:
```diff
@@ -0,0 +1,56 @@
+    @Override
+    public boolean isActivated() {
+        JenkinsLocationConfiguration loc = JenkinsLocationConfiguration.get();
+        return loc == null || loc.getUrl() == null;
```
#2935 forces the null check (so #3021 does not save us here), but if it _is_ null, we should just do nothing because Jenkins has not even started up yet anyway. Thus

```java
return loc != null && loc.getUrl() == null;
```

Wadeck replied:
thank you for the information

jglick commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -227,6 +228,19 @@ public boolean isUsingSecurityToken() {
+        Jenkins j = Jenkins.getInstance();
+        j.checkPermission(Jenkins.ADMINISTER);
+
+        String rootUrl = req.getParameter("rootUrl");
```
Just replace the current method parameters with

```java
@QueryParameter String rootUrl
```

Wadeck replied:
nice feature 👍

jglick commented on core/src/main/resources/jenkins/install/SetupWizard/setupWizardFirstUser.jelly:
```diff
@@ -44,6 +51,29 @@
             $('username').focus();
           </script>
+       </form>
+       <form action="${rootURL}/setupWizard/configureRootUrl" class="root-url no-json" method="post">
```
`saveRootUrl` specifies the URL to post to, so I suspect `action` and `method` could just be deleted.

Wadeck replied:
removed

jglick commented on core/src/main/resources/jenkins/install/SetupWizard/setupWizardFirstUser.properties:
```diff
@@ -1 +1,3 @@
\ No newline at end of file
+Create\ First\ Admin\ User=Create First Admin User
+Other\ Configuration=Other Configuration
+Jenkins\ URL=Jenkins URL
```
trailing newlines please

Wadeck replied:
corrected (also for FR translation)

jglick commented on test/src/test/java/jenkins/diagnostics/RootUrlNotSetMonitorTest.java:
```diff
@@ -0,0 +1,60 @@
+    @Test
+    @Issue("JENKINS-31661")
+    public void testWithRootUrl_notConfigured() {
+        JenkinsLocationConfiguration.get().setUrl(null);
```
Could be merged into the preceding test, save a Jenkins startup.

Wadeck replied:
done

oleg-nenashev reviewed (dismissed):
I do not see any obvious issues in the code

oleg-nenashev commented:
@reviewbybees done

oleg-nenashev commented:
The test failure is unrelated (my hall of shame). 
@daniel-beck are you fine with the testing/fix results?

oleg-nenashev commented:
seems @reviewbybees done

Wadeck committed:
- modification requested during PR by Jesse

kzantow commented on war/src/main/js/pluginSetupWizardGui.js:
```diff
@@ -905,16 +905,25 @@ var createPluginSetupWizard = function(appendTarget) {
 	};
-	
+
+	var beforeFirstUserSendRootUrl = function(callback){
```
I'd rename this to `saveJenkinsUrl` or something

kzantow reviewed (approved):
SGTM :+1:

Wadeck committed:
- renaming proposed by Keith

daniel-beck commented:
I'd *really* like to take a look at how this works, so on-holding this for a bit longer.

oleg-nenashev commented:
@daniel-beck any updates / ETAs? No urgency though. The LTS baseline has passed though it may be backportable to .2/.3.

oleg-nenashev commented:
@daniel-beck gentle ping

daniel-beck commented:
This would also take care of [JENKINS-28466](https://issues.jenkins-ci.org/browse/JENKINS-28466).

oleg-nenashev commented:
IMO it's ready to go. The task has been waiting for @daniel-beck 's review for 2 months, we should do something about that. I will try to sync-up with him next week

daniel-beck commented:
This option hijacks the 'Create admin user' screen for a second purpose. I'm not convinced this is great (a separate screen would probably be cleaner), but not strongly opposed either. Both 'Save' and 'Continue as admin' apply the setting that is entered, and it uses the URL the user accesses Jenkins with as the default (similar to the global config screen).

The 'Manage Jenkins' screen shows a failing reverse proxy monitor with misleading error message if the user enters a different URL than the one they use Jenkins with. Ideally we adapt the error message to include mention of a different configured URL than it is accessed with, if we haven't already (I'm still on the PR build)

Skipping the setup wizard continues to not set a location. Unsure whether this is preferable to potentially setting a wrong/temporary one when skipping. Given that 'Skip admin user' saves the specified URL even if unchanged, this seems inconsistent at least.

Thoughts?

Wadeck commented:
> a separate screen would probably be cleaner

It was my initial plan but due to 
- the complexity to add a new page in the process
- the load increase in terms of UX (new page of a form = more user skip it)
- the quite straightforward process to fill that additional information

I was not totally convinced the new page worth it.

> Skipping the setup wizard continues to not set a location. [...]

In the case the user wants to setup the URL but skip the admin creation, we need to add a new button ? Or proposing a check box on the left of the root URL saying something like "Do you want to set the root URL now ?". Or a save button just on the right with a nice icon to only save the root URL ?

I am a bit balanced between the new page approach and the current one since they are two opposite solution.

Any thoughts @kzantow ? What would you have done if this was a requirement since the beginning ?

oleg-nenashev commented:
Not ready for merge. @daniel-beck and @Wadeck need to agree on the PR first

kzantow commented:
@daniel-beck @Wadeck I agree with Daniel that hijacking the wrong place for this. Is the idea that we want to **a**) force the user to enter something? or b) have them accept the default with the option to change it?

If it's **a**, then we should add an empty text box and force them to fill it in on a separate step, probably a "Configure Jenkins" step, which could be extended with additional configuration as needed.

If it's **b**, we should just not add this to the installer UI and not add an administrative monitor at all.

daniel-beck commented:
Setting the location of the UI controls aside, when a user skips the wizard, we could force set it to the current value. In some situations that'll result in invalid values (e.g. setting up via localhost on Windows server, or using an SSH tunnel before exposing to outside), but it's probably not worse than having it not set (unsure about that though).

kzantow commented:
It might be a good idea to add a "Configure Jenkins" step that people can contribute other necessary configurations for, depending on the plugins they install. It could be skipped if there are no "necessary" configurations.

To me this is a "good idea" configuration step, which would preclude using the admin monitor UNLESS you install some plugins that need it, like Github PR webhook callbacks, in this case it's necessary and adding a "Configure Jenkins" step that the user can't skip would be appropriate.

daniel-beck commented:
@kzantow Not saying it's not, but right now it's possible to skip all the config at the start at any point, and you get no plugins, and the default admin user. I'm thinking about what the URL should look like in such cases.

kzantow commented:
@daniel-beck maybe setting it to any non-localhost value would be "good enough"

daniel-beck commented:
@kzantow If you access it as localhost, there's no useful other default we could use. Between reverse proxies and multiple network interfaces,we have no way to know what's the canonical URL.

kzantow commented:
@daniel-beck right, i mean, if you access it and the request has a host value that's NOT localhost, we could default it to that value. In the case of localhost, we could use some `getHostName` lookup on the interfaces, but none of these things would necessarily be accurate; actually they will almost always be wrong, as presumably the Jenkins instance is behind a firewall and there's some forwarding from an outward-facing router which we don't know anything about. So let's not ever default this value to anything. I'm pretty sure that's what you're saying in the first place. :)

I guess, the point is, there is a case where some plugin needs some additional configuration to function properly (e.g. getting Github webhooks) and we could try to prompt the user in that case, but as you said, the user could skip the installer or parts of it, potentially. There should be some pluggable mechanism to contribute these required configurations, but this PR is not that thing. I suggest that we make an extension point for this to contribute both to the setup wizard and to admin monitors when these things are not properly set up as the correct solution for the problem. I assume this would probably be a low priority thing, though and suggest deferring until later.

Wadeck commented:
I proposed to close this PR by non-merging and using the new ticket : [JENKINS-49287](https://issues.jenkins-ci.org/browse/JENKINS-49287) for the new story concerning the custom wizard page.

WDYT @kzantow @daniel-beck ?

daniel-beck commented:
The new wizard page is out of scope for this change. I don't consider that important enough to block a proper implementation of this improvement.

Wadeck committed:
- rework to have a new page in the SetupWizard for the root URL

Wadeck commented:
I added a new page to provide the possibility to enter the Jenkins root URL after the admin creation.

During admin creation, the user can skip or save the admin user, in both case he then goes to the "Instance configuration" page.

At that step, the root URL default value correspond to the request base url but can be modified. The user can skip or save.

Accordingly to the skip actions he takes, the wizard will display 0-1-2 warning message about its skip(s).

<details>
<summary>Screenshot</summary>

![screenshot_2018-02-09_194334_001](https://user-images.githubusercontent.com/2662497/36045034-b927a51c-0dd4-11e8-9742-e1c8aeb1f894.png)

PS: due to some weird behavior in terms of language management, I cannot set the language to English for the wizard (but can for the rest of the application by simply using a small chrome extension). Due to a different mechanism of language selection
</details>

I keep the instance configuration page within an iframe in order to ease the integration of some extension point like discussed previously.

Just as a reminder, if the user skips the admin creation, he will be able to use the admin:<randomPwd> like before and in case he skips the root URL part, the admin monitor will show up. Skipping entire wizard will "trigger" both behavior like the current version.

oleg-nenashev commented:
Merge conflict. It would be nice to get this PR landed tho

Wadeck committed:
Merge branch 'master' into JENKINS-31661_ROOT_URL_NOT_CONFIGURED

Wadeck commented:
Note to myself, need to merge after the recent InstallState serialization bug fix

Wadeck committed:
Merge branch 'master' into JENKINS-31661_ROOT_URL_NOT_CONFIGURED

daniel-beck commented:
Build failure looks relevant.

Wadeck committed:
Correct relevant non-conflicting merge changes

Wadeck commented:
Come from a merge inside Wizard that was not conflicting but changing function name. Same modifications is now applied to this PR to be consistent.

daniel-beck commented:
I tested PR build 14. Another nice improvement, glad to see the new screen.

Regarding the wording about why the URL needs to be configured: While I think there may be problems related to Stapler, I don't know whether they occur when the URL is not set, or when the URL is different from the actually accessed one.

An easy(ish) to understand example would be email notifications, or PR build statuses: Jenkins will need to know its URL to properly publish links to it in these situations (when it cannot infer the URL from the request).

### Behavior notes

* It is slightly weird to see it still show the new system configuration form when closing the wizard via X. I'd have expected the "Setup complete" screen. Not a blocker.
* I got `WARNING: JENKINS-49795: attempt to (de-)serialize anonymous class jenkins.install.InstallState$7` logged. This might need adapting to the recent changes to `InstallState` classes. CC @jglick 

### Presentation notes

**New configuration page:**

* Inconsistency: "Create First Admin User" is title case, "System configuration" is sentence case
* No specific explanation why the Jenkins URL needs configuring (or what to even configure here in case of multiple possibilities etc.) on the new config screen. While there's no explanation for the admin user, that seems like configuration more easily understood, as opposed to the Jenkins URL configuration.
* Labels are confusing: "Terminate" should be "finish". In English, it's more associated with operations like "Abort".
* It's unclear to users what the difference between setting to localhost and skipping it is.

**"Jenkins is ready" page:**

* Double single quote in message: "You''ve skipped" should be "You have skipped"
* "You''ve skipped the instance configuration. " is unnecessarily vague (unless that page is extensible and/or we expect additions in the near future.)
* "You can still go in "Manage Jenkins" pages to fill the forms." should be more specific, e.g. "To configure the URL of Jenkins, go to …"

**Admin monitor:**

* Wording should be made more specific, if we can provide clear examples. `BUILD_URL` env vars (JENKINS-28466), links to Jenkins in build failure emails, or PR statuses should be easy to understand.

Wadeck committed:
- in case of skip => also skip the instance configuration
- terminate => finish (+ uppercase)
- replace double single quotes
- add help message in the wizard to inform admins
- correct the installState serialization problem
- update working (admin monitor + wizard)

Wadeck commented:
@daniel-beck thank you for the detailed review. I corrected all the points. 
Initially I let some vague explanation to allow future extensions but as it's not a prority, this could lead to confusion.

Monitor re-wording:
![root_url_wizard_monitor](https://user-images.githubusercontent.com/2662497/37594730-59791e4c-2b76-11e8-8879-1495c20a240d.png)

Wizard help text:
![root_url_wizard_help](https://user-images.githubusercontent.com/2662497/37594731-59a72ba2-2b76-11e8-9488-c0c147076f0c.png)

daniel-beck reviewed (changes requested):
I tested build 17.

Findings that should be addressed:

* Clicking X on the first form will now directly skip to the last screen, but only show the note about the root URL, not the note about the user (which is the more important one). If only one note is shown here, it should be the admin user one. Skipping both user and URL config individually shows both notes on the last screen, but X'ing only shows the URL note. With the admin monitor, the note on skipping that config is probably redundant, and we could limit it to just the admin user note (required so people know how to log in)?
* The label "Skip the step and terminate" still exists. Perhaps just "Not now"?

Other notes:

* With the new subtitle, the layout of the form looks slightly weird.

Otherwise looks good 👍 

Perhaps @rtyler or @bitwiseman want to do some wordsmithing?

Wadeck committed:
address Daniel's comment
- skip all steps will show two messages instead of one (like before when clicking on skip twice)
- skip => not now
- remove margin and adjust padding for the wizard forms
- add the "if possible"

Wadeck commented:
@daniel-beck comments addressed + screenshots:

In case of skip:
![wizard_double_skip_message](https://user-images.githubusercontent.com/2662497/37650748-1ae5a554-2c36-11e8-82b0-a6f0b76c0841.png)

The form is centered in terms of padding now, so less weird I think, are you ok ?
![better_padding_remove_margin](https://user-images.githubusercontent.com/2662497/37650716-ffda3a68-2c35-11e8-889a-568d63e11464.png)

bitwiseman commented:
@daniel-beck What did you want smithed?

daniel-beck commented:
@bitwiseman An addition to the setup wizard, after admin user creation it now has a new page to set the Jenkins URL. Localized strings are distributed across multiple `*.properties` files in this PR.

I remember @rtyler substantially improved the original 2.0 setup wizard, so I thought I'd ping you for this.

aheritier commented:
@Wadeck I think in french we use more, in such a case, the verb "passer" instead of "sauter" for "to skip". But if "sauter" is already used.... let's go

Wadeck committed:
address Arnaud's comment on the sauter vs passer

Wadeck commented:
@aheritier c'est corrigé. Good point there were multiple usage of both so I make that consistent.

bitwiseman reviewed (commented):
Suggested changes to wording.

Wadeck commented on core/src/main/resources/jenkins/diagnostics/RootUrlNotSetMonitor/message.properties:
```diff
@@ -0,0 +1,3 @@
+blurb=Jenkins root URL is required for the correct operation of many Jenkins features like email notifications, \
```
Provide instead of provided no ?

daniel-beck replied:
@Wadeck Yes, typo.

Wadeck commented on core/src/main/resources/jenkins/install/SetupWizard/setupWizardConfigureInstance.properties:
```diff
@@ -0,0 +1,6 @@
+Instance\ Configuration=Instance Configuration
+Jenkins\ URL=Jenkins URL
+jenkinsURL_help=The Jenkins URL is used to provide absolute links for different external usages like \
```
to provide as the root URL, never heard that wording, you confirm it's not a typo ?

dwnusbaum replied:
Looks like a typo to me. A possible change is to delete `as`: `to provide as the root URL` -> `to provide the root URL`

bitwiseman replied:
@dwnusbaum  is correct. I've edited the comment to match.

Wadeck committed:
address Liam's comment

Wadeck committed:
correct previous commit typo

Wadeck committed:
correct previous commit typo (2)

Wadeck committed:
add blank end line

jtnord commented on core/src/main/java/jenkins/model/JenkinsLocationConfiguration.java:
```diff
@@ -50,6 +50,17 @@
+    public static @Nonnull JenkinsLocationConfiguration getOrDie(){
+        JenkinsLocationConfiguration config = JenkinsLocationConfiguration.get();
+        if (config == null) {
+            throw new IllegalStateException("JenkinsLocationConfiguration instance is missing. Probably the Jenkins instance is fully loaded at that time.");
```
typo?  "Probably the Jenkins instance is fully loaded at that time"

Don't you mean **not** fully loaded  (also should be this time)

Wadeck committed:
address James' comment

Wadeck committed:
Merge remote-tracking branch 'origin/JENKINS-31661_ROOT_URL_NOT_CONFIGURED' into JENKINS-31661_ROOT_URL_NOT_CONFIGURED

jtnord reviewed (commented):
LGTM

daniel-beck commented:
Skipping the admin user creation but not skipping the root URL configuration results in a "Finished" page without any notes, when it should show the "Admin User" note.

daniel-beck reviewed (dismissed):
I tested PR merge build 8:

* Skipping the admin user creation but not skipping the root URL configuration results in a "Finished" page without any notes, when it should show the "Admin User" note.

Wadeck commented:
@daniel-beck normally the lines [917](https://github.com/jenkinsci/jenkins/pull/3082/files#diff-1b14e0dbd9ab104cdbcea2c1f823bcacR917) and [961](https://github.com/jenkinsci/jenkins/pull/3082/files#diff-1b14e0dbd9ab104cdbcea2c1f823bcacR961) are meant to do that job. I will investigate this point.

daniel-beck commented:
@Wadeck As I wrote before, I'm not sure the 'root URL' message is particularly useful given the existence of the monitor, so if it's easier to just not have that one, should be fine.

Wadeck committed:
address Daniel's comment

Wadeck commented:
@daniel-beck bug corrected, was introduced by an invalid merge from my side.

dwnusbaum commented on core/src/main/java/jenkins/install/InstallState.java:
```diff
@@ -56,6 +57,7 @@
         new InstallState("UNKNOWN", false) {},
         new InstallState("INITIAL_SETUP_COMPLETED", false) {},
         new InstallState("CREATE_ADMIN_USER", false) {},
+        new InstallState("CONFIGURE_INSTANCE", false) {},
```
I think that adding to this list will break deserialization from old Jenkins versions. The other items are here in the order in which they occurred in the source when they were anonymous classes, so inserting into the list will break the migration by causing any instances whose serialized state was `jenkins.install.InstallState$4`, which corresponds to `INITIAL_SECURITY_SETUP`, to be converted into the `CONFIGURE_INSTANCE` state when deserialized, which is probably not intentional.

Since no old instances will have this as the serialized state, I think you are fine with just the named class below and this line should be removed. Maybe also good to add something like "DO NOT ADD TO THIS LIST" to its Javadoc.

dwnusbaum commented on core/src/main/java/jenkins/model/JenkinsLocationConfiguration.java:
```diff
@@ -50,6 +50,17 @@
+    /**
+     * Gets local configuration. For explanation when it could die, see {@link #get()}
+     */
+    public static @Nonnull JenkinsLocationConfiguration getOrDie(){
```
Probably should add `@Restricted(NoExternalUse.class)` since it is a new public method.

dwnusbaum commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -293,6 +296,24 @@ public HttpResponse doCreateAdminUser(StaplerRequest req, StaplerResponse rsp) t
     }
+    
+    @RequirePOST
+    public HttpResponse doConfigureRootUrl(StaplerRequest req, @QueryParameter String rootUrl) {
```
Probably should add @Restricted(NoExternalUse.class) since it is a new public method.

dwnusbaum commented on core/src/main/java/jenkins/install/SetupWizard.java:
```diff
@@ -293,6 +296,24 @@ public HttpResponse doCreateAdminUser(StaplerRequest req, StaplerResponse rsp) t
+        
+        LOGGER.log(Level.FINE, "Root URL set during SetupWizard to {0}", new Object[]{ rootUrl });
+        
+        JenkinsLocationConfiguration.getOrDie().setUrl(rootUrl);
```
nit: Could be nice to try parsing the URL and return an error if it isn't a valid URL. (You could also do the same thing in the admin monitor and have the message state that the URL appears to be invalid.)

dwnusbaum reviewed (changes requested):
I think that adding to `InstallState.UNUSED_INNER_CLASSES` is a bug and should be removed. Other than that I just saw 2 restricted annotations that would be good to add.

I ran through the setup wizard locally with the PR and tried various combinations of skipping steps and everything worked ok. The admin monitor also worked fine for me, so once the above changes are made I am happy to approve it.

Wadeck committed:
address Devin's comment
- implement error management in the wizard
- provide different message in the monitor in case of null / invalid url
- add an helper to validate URL (even with _ in domain)
- add @Restricted

Wadeck commented:
To check the URL was a bit more tricky that expected due to the totally different nature of the Wizard, but now it's done.

<details><summary>URL validation behavior (gif)</summary>

![wizard-ajax-error](https://user-images.githubusercontent.com/2662497/38084773-6f22e338-334e-11e8-8876-5a7870a43dce.gif)

</details>

<details><summary>Monitor message relative to the null/invalid URL (gif)</summary>

![wizard-monitor-different](https://user-images.githubusercontent.com/2662497/38084789-81aa9690-334e-11e8-86c3-9eecccebbb4a.gif)

</details>

Wadeck committed:
- also update the monitor test to be consistent with its new behavior

dwnusbaum reviewed (approved):
Small comments about some of the wording but looks fine to me in general.

I am somewhat against adding a new dependency but if `new URL(input).toURI()` is totally inadequate then I am ok with it.

Wadeck commented on core/src/main/java/jenkins/util/UrlHelper.java:
```diff
@@ -0,0 +1,55 @@
+import java.net.URL;
+
+public class UrlHelper {
+    public static boolean isValid(String url){
```
Annotation added. 

For the second point, actually only the Wizard and the admin monitors are asking for valid URL. But in case you have a very weird URL that is not considered valid for this helper, you can still use it in the configuration page, no check are done there. You will "just" have to disable the admin monitor.

Wadeck committed:
address Devin's comments

Wadeck commented on core/src/main/java/hudson/util/HttpResponses.java:
```diff
@@ -89,16 +89,28 @@ public static HttpResponse okJSON(@Nonnull Map<?,?> data) {
+     *
+     * @since TODO
+     */
+    public static HttpResponse errorJSON(@Nonnull String message, @Nonnull Map<?,?> data) {
```
See #3379

Wadeck commented on core/src/main/resources/jenkins/install/SetupWizard/authenticate-security-token.jelly:
```diff
@@ -1,3 +1,26 @@
+<!--
+The MIT License
+
+Copyright (c) 2018, CloudBees, Inc.
```
I think I added the header to the wrong file -.-

Wadeck committed:
correct header files

kuisathaverat reviewed (approved):
LGTM

Wadeck committed:
include relevant classes for the url validation from commons-validator:1.6

Wadeck commented on core/pom.xml:
```diff
@@ -309,6 +309,17 @@ THE SOFTWARE.
+    <dependency>
+      <groupId>commons-validator</groupId>
+      <artifactId>commons-validator</artifactId>
+      <version>1.6</version>
```
👍 addressed, 4 classes included with `@Restricted`

Wadeck committed:
Move Apache classes in a more standard package

dwnusbaum reviewed (approved):
Looks good to me. Maybe the imported classes could be moved into a different package since `jenkins.org.blah` is a little weird but I don't feel strongly about it.

Wadeck commented:
> Maybe the imported classes could be moved into a different package...
@dwnusbaum it was based on the same concept as used in [hudson/org/apache/tools/tar/TarInputStream.java](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/org/apache/tools/tar/TarInputStream.java)

Wadeck commented:
@daniel-beck are we good with this one ?

daniel-beck commented:
To be merged later this week.

dwnusbaum commented:
@reviewbybees done

jglick commented on core/src/main/resources/jenkins/install/SetupWizard/setupWizardConfigureInstance.jelly:
```diff
@@ -0,0 +1,122 @@
+            $('root-url').focus();
+            (function setInitialRootUrlFieldValue(){
+              var iframeUrl = window.location.href;
+              var iframeRelativeUrl = '/setupWizard/setupWizardConfigureInstance';
```
Wrong! This strips the trailing slash, which is required. [JENKINS-51660](https://issues.jenkins-ci.org/browse/JENKINS-51660)

oleg-nenashev commented:
@jglick a follow-up PR would be much appreciated
